// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless  by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

syntax="proto3";
package ocean_kv.data;

option cc_generic_services = true;


message IndexInfo {
       uint32 file_id = 1; // key对应value所在文件id
       uint32 total_length = 2;
       uint32 head_length = 3;
       uint64 shift = 4;
}

message GetDataRequest {
       string key = 1;
       IndexInfo index_info = 2;
};

message GetDataResponse {
       uint32 code = 1;
       bytes data = 2;
};

message PutDataRequest {
       string key = 1;
};

message PutDataResponse {
       uint32 code = 1;
       IndexInfo index_info = 2;
};

message DeleteDataRequest {
       string key = 1;
       IndexInfo index_info = 2;
       bool need_data = 3;
};

message DeleteDataResponse {
       uint32 code = 1;
       bytes data = 2;
};

service DataService {
      rpc GetData(GetDataRequest) returns (GetDataResponse);
      rpc PutData(PutDataRequest) returns (PutDataResponse);
      rpc DeleteData(DeleteDataRequest) returns (DeleteDataResponse);
};
